src/query/plan/rewrite/periodic_delete.hpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <memory>
+#include "query/plan/operator.hpp"
+
+namespace memgraph::query::plan {
+
+namespace impl {
+
+template <class TDbAccessor>
+class PeriodicDeleteRewriter final : public HierarchicalLogicalOperatorVisitor {
+ public:
+  PeriodicDeleteRewriter(SymbolTable *symbolTable, AstStorage *astStorage, TDbAccessor *db)
+      : symbol_table(symbolTable), ast_storage(astStorage), db(db) {}
+
+  ~PeriodicDeleteRewriter() override = default;
+
+  using HierarchicalLogicalOperatorVisitor::PostVisit;
+  using HierarchicalLogicalOperatorVisitor::PreVisit;
+  using HierarchicalLogicalOperatorVisitor::Visit;
+
+  bool Visit(Once &t) override { return true; }
+
+  bool PreVisit(PeriodicCommit &op) override {
+    commit_frequency_ = op.commit_frequency_;
+    return true;
+  }
+
+  bool PreVisit(PeriodicSubquery &op) override {
+    periodic_subquery_seen_ = true;
+    return true;
+  }
+
+  bool PreVisit(Delete &op) override {
+    // since periodic commit needs to flush deltas over time, buffer size needs to be put in delete
+    // so it can know when to trigger delete of nodes and/or relationships
+    if (commit_frequency_) {
+      op.buffer_size_ = commit_frequency_;
+    }
+    if (periodic_subquery_seen_) {
+      throw utils::NotYetImplemented("Delete can not be used with periodic subquery, please use USING PERIODIC COMMIT");
```

Comment:
Consider these 2 plans
This one we can do now
```
memgraph> explain using periodic commit 2 match (n) detach delete n;
+---------------------+
| QUERY PLAN          |
+---------------------+
| " * EmptyResult"    |
| " * PeriodicCommit" |
| " * Delete"         |
| " * ScanAll (n)"    |
| " * Once"           |
+---------------------+
```
We can do this plan because periodic commit is in the same branch as delete, and we can transfer easily the information about what is going to be the commit frequency, so periodic commit and delete can be in sync by how often we flush and periodically commit (after 2 rows, both delete will perform the exact delete and the periodic commit will commit).

We need this because Delete is currently a pipeline breaker, it will not delete until it has aggregated everything it can delete, so delete operations can be efficient. If it does not delete in that manner, deletes will take n^2 time if nodes are detach deleted one by one.

The second plan is:
```
memgraph> explain match (n) CALL { WITH n detach delete n } in transactions of 2 rows;
+-----------------------+
| QUERY PLAN            |
+-----------------------+
| " * EmptyResult"      |
| " * Accumulate"       |
| " * PeriodicSubquery" |
| " |\\ "               |
| " | * EmptyResult"    |
| " | * Delete"         |
| " | * Produce {n}"    |
| " | * Once"           |
| " * ScanAll (n)"      |
| " * Once"             |
+-----------------------+
```
We can't execute this because currently, how we execute subqueries, is that we exhaust the subquery branch when we do an input pull. Since the first operator in the subquery branch (right one) is `Once`, we will pull `Once` and the second time it will signal the `Delete` to execute, since once returned false. Therefore, deletes will take `O(n^2)` complexity.
Idea therefore is to introduce some other kind of operator because `Delete` behaves different from other operators, but since we have a way to periodically execute deletes, I would postpone that work.

Reply:
Thank you for explanation! I would suggest to add more obvious error message what user should do in case of this error message. Because, if me as user got this message, I still won't know what to do and will need to go to the documentation and try to understand what developers wanted from me :D